Add `qutebrowser-consult.el`

minimaximalist opened the pull request:
We discussed adding a new `qutebrowser-consult.el` and putting the remaining Consult related code from `qutebrowser.el` there in https://github.com/lrustand/qutebrowser.el/issues/18. This PR adds a rough draft of `qutebrower-consult.el`.

lrustand commented on qutebrowser-consult.el:
```diff
@@ -0,0 +1,238 @@
+      (qutebrowser (car selected)))))
+
+;;;###autoload
+(defun qutebrowser-consult-launcher-tab (&optional initial)
+  "Select a URL to open in a new tab using Consult.
+
+Set initial completion input to INITIAL.
+Modify `qutebrowser-consult-launcher-sources' to change which sources
+are included."
+  (interactive)
+  (qutebrowser-consult-launcher initial 'tab))
+
+;;;###autoload
+(defun qutebrowser-consult-launcher-window (&optional initial)
+  "Select a URL to open in a new window using Consult.
+
+Set initial completion input to INITIAL.
+Modify `qutebrowser-consult-launcher-sources' to change which sources
+are included."
+  (interactive)
+  (qutebrowser-consult-launcher initial 'window))
+
+;;;###autoload
+(defun qutebrowser-consult-launcher-private (&optional initial)
+  "Select a URL to open in a private window using Consult.
```
I don't think we need to duplicate all of these launcher functions.

I think it would be better to add a defcustom in the `qutebrowser.el` file to select which launcher backend should be used for the various launcher functions. Similar to how it is done for `qutebrowser-command-backend`.

The variable could be called something like `qutebrowser-select-url-backend` or `qutebrowser-launcher-backend` depending on how it is implemented.

lrustand commented on qutebrowser-consult.el:
```diff
@@ -0,0 +1,238 @@
+	 (qutebrowser-default-open-target (or target qutebrowser-default-open-target))
+	 (selected (consult--multi
+		    (mapcar #'qutebrowser-consult--shorten-source-name
+			    (mapcar #'eval qutebrowser-consult-launcher-sources))
```
Why eval here?

minimaximalist replied:
`qutebrowser-consult--shorten-source-name` doesn't work on symbols, it needs to be passed the source plist. I'm fine eliminating it entirely; my goal with it was to make the Consult-based launcher appear similar to the normal one but avoid confusion when the Consult sources are used elsewhere.

lrustand replied:
Yeah, I think it would be best to eliminate `qutebrowser-consult--shorten-source-name`. Even though the Qutebrowser prefix is redundant when used specifically through the launcher, I don't think it does any harm to keep the prefix.

lrustand commented on qutebrowser-consult.el:
```diff
@@ -0,0 +1,238 @@
+  (interactive)
+  (qutebrowser-consult-launcher initial 'private-window))
+
+(define-minor-mode qutebrowser-consult-mode
+  "Use Consult for Qutebrowser completion."
+  :lighter nil
+  :global t
+  (if qutebrowser-consult-mode
+      (progn
+	(advice-add 'qutebrowser-launcher :override #'qutebrowser-consult-launcher)
+	(advice-add 'qutebrowser-launcher-tab :override #'qutebrowser-consult-launcher-tab)
+	(advice-add 'qutebrowser-launcher-window :override #'qutebrowser-consult-launcher-window)
+	(advice-add 'qutebrowser-launcher-private :override #'qutebrowser-consult-launcher-private))
+    (advice-remove 'qutebrowser-launcher #'qutebrowser-consult-launcher)
+    (advice-remove 'qutebrowser-launcher-tab #'qutebrowser-consult-launcher-tab)
+    (advice-remove 'qutebrowser-launcher-window #'qutebrowser-consult-launcher-window)
+    (advice-remove 'qutebrowser-launcher-private #'qutebrowser-consult-launcher-private)))
```
This is actually also a quite nice way to handle it. But I think I still prefer the approach I described above.

lrustand commented on qutebrowser-consult.el:
```diff
@@ -0,0 +1,238 @@
+;;; qutebrowser-consult.el --- Consult completion for Qutebrowser -*- lexical-binding: t; -*-
+
+;; Copyright (C) 2024 Isaac Haller & Lars Rustand.
```
Did you not update your calendar yet? ;)

minimaximalist replied:
Oops lol

lrustand commented on qutebrowser-consult.el:
```diff
@@ -0,0 +1,238 @@
+	:category 'other
+	:action #'qutebrowser-send-commands
+        :annotate #'qutebrowser-consult--annotate
+	:async
+	(consult--dynamic-collection
+	    (lambda (input)
+	      (qutebrowser-command-search (string-split (or input ""))))
+	  :min-input 0
```
I don't think this should be async, since the `qutebrowser-command-search` function retrieves all the candidates every time it is called, so it doesn't benefit from being async. In fact it would be much more performant as a static source, since then it would only need to fetch the command list once.

lrustand replied:
The only reason it was used asynchronously previously is because it wasn't possible to combine dynamic and static.

minimaximalist replied:
Is there a hook in `qutebrowser.el` that could be used to initialize the list of candidates? Ideally one that runs when RPC connects or something like that.

lrustand replied:
Not a hook per se, but you could probably add it to `qutebrowser-populate-commands`.

minimaximalist replied:
I've modified the source to be static. It calls `qutebrowser-command-search` once per run which is a lot better than before and cleaner than keeping a list of candidates in a variable.

lrustand commented on qutebrowser-consult.el:
```diff
@@ -0,0 +1,238 @@
+	:action #'qutebrowser-open-url
+	:annotate #'qutebrowser-consult--annotate
+	:async
+	(consult--dynamic-collection
+	    (lambda (input)
+	      (qutebrowser--history-search (string-split (or input ""))
+					   qutebrowser-dynamic-results))
+	  :min-input 0
```
I think Consult's async completion has some sort of builtin string split functionality, but I've never looked into it in much detail. Do you think this is something we could leverage here?

lrustand replied:
If you don't know either, feel free to ignore this comment.

minimaximalist replied:
I don't think that functionality does what we want in this situation; it's related to `consult-async-split-style`.

lrustand reviewed (changes requested):
Thanks, this is great!

Left a few comments, but overall this looks very good. I haven't tested it yet, but will report back later after I test it.

lrustand commented on qutebrowser-consult.el:
```diff
@@ -0,0 +1,240 @@
+	:history nil
+	:category 'url
+	:action #'qutebrowser-open-url
+	:new #'qutebrowser-open-url
```
Has Consult been updated to handle this correctly now? Last I checked it only worked for the first source in the list. https://github.com/minad/consult/issues/1088

minimaximalist replied:
It does work when the source has been narrowed to. I add this source to `consult-buffer` in my setup and without that line trying to start web searches creates plain buffers instead.

minimaximalist commented on qutebrowser-consult.el:
```diff
@@ -0,0 +1,238 @@
+  :prefix "qutebrowser-consult")
+
+;;;; Helper functions
+(defun qutebrowser-consult--format-entry (entry)
+  "Modify ENTRY for presentation."
+  (qutebrowser--shorten-display-url (qutebrowser--tofu-strip entry)))
```
> Adding tofus to the candidates and then removing them again before passing the candidates to Consult feels messy, and I think we should instead modify the various -search functions to not add the tofus in the first place. The completing-read backend still needs tofus for dispatching though, so a small wrapper function would be needed there.

Agreed. I originally didn't want to touch the existing completion code too much, but that would be cleaner.

minimaximalist commented on qutebrowser-consult.el:
```diff
@@ -0,0 +1,238 @@
+'Qutebrowser' is stripped from the front and the remaining first letter is capitalized."
+  (let ((new-source (seq-copy source))
+	(new-name (string-remove-prefix "Qutebrowser " (plist-get source :name))))
+    (plist-put new-source :name (capitalize new-name))))
```
I've removed this function and calls to it.

lrustand commented on qutebrowser-consult.el:
```diff
@@ -0,0 +1,222 @@
+  (let* ((consult-async-min-input 0)
+	 (consult-async-split-style nil)
+	 (selected
+	  (consult--multi `(,(plist-put (seq-copy qutebrowser-consult--command-source)
+					:name nil))
```
```suggestion
	  (consult--multi '(qutebrowser-consult--command-source)
                          :group nil
```
This can be simplified by overriding the `:group` function in `consult--multi`.

minimaximalist commented on qutebrowser-consult.el:
```diff
@@ -0,0 +1,238 @@
+  (let ((new-source (seq-copy source)))
+    (plist-put new-source :action nil)))
+
+;; NOTE does this still need to exist?
+(defun qutebrowser-consult-select-url (&optional initial default)
+  "Dynamically select a URL, buffer, or command using consult.
```
A few functions besides the launcher `qutebrowser-select-url`, so if the completion backend is made selectable I think it makes sense to keep a Consult version of that function.

minimaximalist commented:
Ok, I've removed the duplicate Consult-based launcher commands and added a customizable variable `qutebrowser-selection-function`. `qutebrowser-select-url` now calls this function, and the previous implementation has been renamed to `qutebrowser-select-url-completing-read`.

`qutebrowser-consult` still has to deal with tofus unfortunately; the `qutebrowser` dispatching function requires them to know to do with the candidate that the selection function returns. My solution was to add a custom `:qb-tofu` parameter to the Consult sources that need it and append it to `qutebrowser-consult-select-url`'s return value.

As an aside, when Emacs 30 comes out we could use additional text properties instead of tofus; let binding a new variable `minibuffer-allow-text-properties` allows this (I run an Emacs built from git and briefly experimented with this).

lrustand commented:
> `qutebrowser-consult` still has to deal with tofus unfortunately; the `qutebrowser` dispatching function requires them to know to do with the candidate that the selection function returns. My solution was to add a custom `:qb-tofu` parameter to the Consult sources that need it and append it to `qutebrowser-consult-select-url`'s return value.

Yeah, I anticipated that there could be some issues like this. I looked at it briefly, but think I have an idea for how to resolve it relatively cleanly without having to do any funny business on the Consult end. Instead of having url selection functions as the main component that are chosen with the variable, the variable could instead choose a launcher function/dispatcher which handle both the selection and dispatching.

So, re-adding the one main `qutebrowser-consult-launcher`, and rearranging the `completing-read` counterpart to have a similar function. The `qutebrowser-launcher-` family of functions could use the launcher/dispatcher functions directly, and I think the `qutebrowser` and `qutebrowser-dwim` commands could also be adapted to use a launcher function instead.

Maybe swapping out `qutebrowser-select-url` in the interactive usage of `qutebrowser` and `qutebrowser-dwim` with the launcher/dispatcher function, and then having a `qutebrowser-open-url` in the function body for when it is called non-interactively.

Using `qutebrowser-dwim` as an example, maybe something like this (untested) could work:

```elisp
(defun qutebrowser-dwim (thing &optional target count)
  (interactive (list (or (and (region-active-p)
                              (filter-buffer-substring
                               (region-beginning)
                               (region-end)))
                         (thing-at-point 'url t)
                         (thing-at-point 'symbol t)
                         (thing-at-point 'word t))
                     (pcase current-prefix-arg
                       ('(4) 'tab)
                       ('(16) 'private-window)
                       (_ nil))
                     (and (numberp current-prefix-arg)
                          current-prefix-arg)))
  (if (and (interactive-p) (not thing))
      (funcall qutebrowser-launcher-backend nil default)
    (qutebrowser-open-url thing target)))
```

I also had a thought cross  my mind if it would be more userfriendly if the variable was a simple true/false flag to enable/disable Consult? I'm a bit torn on this one though, since although it might be slightly clearer/more understandable from a user perspective, it is also less flexible.

On a side note, I noticed when looking at these functions again that the `count` argument is not actually used, even though the numeric prefix does work correctly when send commands. So it turns out that the `qutebrowser-rpc-commands` function reads the numeric prefix directly without needing these functions to pass it along. The count argument should probably be removed to avoid confusion.

> As an aside, when Emacs 30 comes out we could use additional text properties instead of tofus; let binding a new variable `minibuffer-allow-text-properties` allows this (I run an Emacs built from git and briefly experimented with this).

I do run Emacs 30, and have also experimented with this, but my results were mixed. If I recall correctly it worked for one of `completing-read` or `consult--multi` but not the other, or something like that, but I might have made some stupid mistake when I tried it.

Using text properties instead of tofus would indeed feel much cleaner, especially since text properties have to be added for annotation purposes already.

Btw, my emacs says that this variable was introduced in version 20.1

minimaximalist commented:
> Instead of having url selection functions as the main component that are chosen with the variable, the variable could instead choose a launcher function/dispatcher which handle both the selection and dispatching.

> So, re-adding the one main qutebrowser-consult-launcher, and rearranging the completing-read counterpart to have a similar function.

I like this approach in theory. The main problem I see with it is that `qutebrowser-delete-from-history` also currently uses `qutebrowser-select-url`. For that command (or any similar future commands), it doesn't make sense for the launcher function to dispatch (open or select a buffer). We could add an argument to the launcher that disables/enables dispatching, making it behave like `qutebrowser-select-url` currently does; what's your opinion?

> I also had a thought cross my mind if it would be more userfriendly if the variable was a simple true/false flag to enable/disable Consult? I'm a bit torn on this one though, since although it might be slightly clearer/more understandable from a user perspective, it is also less flexible.

I think the current method is clear enough; the available options are selectable with Custom and realistically users of this package are probably well-versed in Emacs anyway given they're using EXWM.

> Using text properties instead of tofus would indeed feel much cleaner, especially since text properties have to be added for annotation purposes already.

> Btw, my emacs says that this variable was introduced in version 20.1

I saw online somewhere saying it wasn't added until version 30, but yes, it's available in earlier versions. I've pushed a commit using text properties instead of tofus that seems to work, but I'd appreciate if it could be tested on earlier versions.

lrustand commented:
Glad to see the tofus are gone, that was a nice little cleanup and I think improves code quality overall.

> I like this approach in theory. The main problem I see with it is that `qutebrowser-delete-from-history` also currently uses `qutebrowser-select-url`. For that command (or any similar future commands), it doesn't make sense for the launcher function to dispatch (open or select a buffer).

I don't think `qutebrowser-delete-from-history` is that important really (and it probably shouldn't even use `qutebrowser-select-url` in the first place, since that returns things that are not valid things to delete from history), and if that comes in the way of other more important things I think it is ok to cripple it, i.e. remove the interactive part of it. It is really just a halfcooked idea that I never finished, the plan was to make it into an [embark](https://github.com/oantolin/embark) action that could be used to delete candidates from inside the launcher, and it wasn't really intended to be used as a standalone interactive command.

> We could add an argument to the launcher that disables/enables dispatching, making it behave like `qutebrowser-select-url` currently does; what's your opinion?

I don't have any strong feelings one way or the other TBH. I'm fine with having things the way they are now in your latest revision, but if you think doing something like this could improve it, and you want to do it, then that's ok too. Or if you want to sacrifice the interactive part of `qutebrowser-delete-from-history` to go the dispatcher route.

> > Btw, my emacs says that this variable was introduced in version 20.1
> 
> I saw online somewhere saying it wasn't added until version 30, but yes, it's available in earlier versions. I've pushed a commit using text properties instead of tofus that seems to work, but I'd appreciate if it could be tested on earlier versions.

I don't really have access to any older Emacsen at the moment, the one in Guix seems to be 30 by default.

lrustand commented on qutebrowser.el:
```diff
@@ -528,8 +535,7 @@ Return up to LIMIT results."
     (mapcar (lambda (row)
               (let* ((url (car row))
                      (title (cadr row)))
-                (propertize (qutebrowser--tofu-append url ?h)
-                            'title title)))
+                (propertize url 'qb-type 'url 'title title)))
```
On the other hand, the prefix does communicate quite clearly that this is not a normal text property. If we want to use a prefix, I think we should make it a little prettier though. How about `qutebrowser-candidate-type` or something like that?

minimaximalist replied:
My preference would be to have a prefix. Should it also be added to the existing `title` property?

I agree that a longer prefix is more clear.

lrustand replied:
Yes, add it to any existing property as well.

lrustand commented:
I noticed that `qutebrowser-consult--exwm-buffer-source` has lost the ability to search in the title of buffers, and only matches against the url. The `qutebrowser-exwm-buffer-search` function handles searching in both url and title, so converting the buffer source into an async source would probably enable that functionality again. But, there are pros and cons of each approach, so what do you think?

lrustand commented on qutebrowser-consult.el:
```diff
@@ -0,0 +1,164 @@
+  "Backend for `qutebrowser-select-url' based on Consult."
+  (let* ((consult-async-min-input 0)
+         (consult-async-split-style nil)
+	 (sources (list qutebrowser-consult--exwm-buffer-source
+			qutebrowser-consult--bookmark-source
+			qutebrowser-consult--history-source
+			qutebrowser-consult--command-source))
```
This should probably use `qutebrowser-consult-launcher-sources`

minimaximalist commented on qutebrowser-consult.el:
```diff
@@ -0,0 +1,164 @@
+  "Consult source for open Qutebrowser windows.")
+
+;;;; Bookmark source
+(defvar qutebrowser-consult--bookmark-source
+  (list :name "Qutebrowser bookmarks"
+        :hidden nil
+        :narrow ?m
+        :history nil
+        :category 'bookmark
+        :face 'consult-bookmark
+        :action #'qutebrowser-bookmark-jump
+	:annotate #'qutebrowser-consult--annotate
+        :items #'qutebrowser-bookmark-search)
+  "Consult source for Qutebrowser bookmarks.")
```
It would be better to have a url-displaying source for the launcher, yeah. I'll add it.

minimaximalist commented on qutebrowser-consult.el:
```diff
@@ -0,0 +1,164 @@
+	      'face 'completions-annotations))
+
+;;;; buffer source
+(defvar qutebrowser-consult--exwm-buffer-source
+  (list :name "Qutebrowser buffers"
+        :hidden nil
+        :narrow ?q
+        :history nil
+        :category 'url
+        :action (lambda (entry)
+		  (switch-to-buffer (get-text-property 0 'qb-buffer entry)))
+        :annotate #'qutebrowser-consult--annotate
+        :items #'qutebrowser-exwm-buffer-search)
+  "Consult source for open Qutebrowser windows.")
```
I had the same issue with URLs not being shortened at first too, but for some reason it started working after restarting Emacs; perhaps byte/native compilation is related. Either way, it's probably safer to do it this way.

minimaximalist commented:
I think combining the dispatch and selection functions is a good idea; it would simplify both completion backends. I don't love the idea of reducing `qutebrowser-delete-from-history`'s functionality, but I can't think of a clean way to restrict it to only selecting history (we'd have to duplicate `qutebrowser--completion-table` bare minimum) so for now let's make it non-interactive and perhaps fix that in the future.

As for the buffer source, I'd rather not make it async. I've pushed a commit that enables matching the title by invisibly appending the title to the candidate. It feels a little hackish but it does work.

lrustand commented:
> I think combining the dispatch and selection functions is a good idea; it would simplify both completion backends. I don't love the idea of reducing `qutebrowser-delete-from-history`'s functionality, but I can't think of a clean way to restrict it to only selecting history (we'd have to duplicate `qutebrowser--completion-table` bare minimum) so for now let's make it non-interactive and perhaps fix that in the future.

I think this is a good decision. `qutebrowser-delete-from-history` is not really convenient to use as an interactive command like that anyway, it would be much more practical as an embark action, or as a more bespoke interactive interface.

I think in the future I might make some stuff like this available only in the Consult extension, in cases where it is troublesome to support both backends.

> As for the buffer source, I'd rather not make it async. I've pushed a commit that enables matching the title by invisibly appending the title to the candidate. It feels a little hackish but it does work.

I agree that it is a bit hackish, but it gets the job done. I don't see any better way of accomplishing it short of going async, so I think this is good enough.

lrustand commented on qutebrowser.el:
```diff
@@ -570,7 +570,7 @@ Return up to LIMIT results."
   (with-current-buffer (or buffer (current-buffer))
     (or qutebrowser-exwm-current-url
         ;; Keep backward compatibility for now
-        (get-text-property 0 'url (buffer-name buffer)))))
+        (get-text-property 0 'qutebrowser-url (buffer-name buffer)))))
```
This one is actually a _different_ `url` text property. It is a remnant from when I used to set a special title format in the Qutebrowser config and parse it in an EXWM hook to add a text property directly to buffer name itself, which was the way the url of a buffer was communicated from QB to EXWM before the implementation of the RPC. I think it is time to ditch this one now, and just rely on `qutebrowser-exwm-current-url`.

minimaximalist replied:
That makes sense to me.

minimaximalist commented:
Ok, hopefully this PR is nearing completion. Candidate selection and dispatching are now done in `qutebrowser-completing-read-launcher` and `qutebrowser-consult-launcher`. `qutebrowser-launcher` now funcalls `qutebrowser-launcher-backend`. `qutebrowser` and `qutebrowser-dwim` have been reworked to use the launcher function when interactive and either send a command or open a URL when noninteractive. For the latter, I've added a user variable (nil by default) that enables sending commands; sending commands is new behavior for this function so we shouldn't surprise users.

lrustand commented:
> Ok, hopefully this PR is nearing completion.

Indeed, it has been a long one, and I'm sorry if I have been too demanding. However, the end result is a very nice solution, and it has even resulted in quite a bit of improvement in the existing code in addition to the added functionality. I really appreciate the effort you've made in this, and I'm very pleased with the outcome!

From what I can see, it looks like everything is now working as expected, and I don't think I have anything more to comment on. So if you are ready to have it merged, I will merge it now.

> Candidate selection and dispatching are now done in `qutebrowser-completing-read-launcher` and `qutebrowser-consult-launcher`. `qutebrowser-launcher` now funcalls `qutebrowser-launcher-backend`. `qutebrowser` and `qutebrowser-dwim` have been reworked to use the launcher function when interactive and either send a command or open a URL when noninteractive. For the latter, I've added a user variable (nil by default) that enables sending commands; sending commands is new behavior for this function so we shouldn't surprise users.

That sounds like a good idea.

lrustand commented:
> So if you are ready to have it merged, I will merge it now.

Actually, scratch that. Now that I approved the changed you should probably be able to merge it yourself.

minimaximalist commented:
I wasn't trying to complain, sorry if it sounded that way. I'm much happier with this PR now than initially; let's merge it.

lrustand commented:
> I wasn't trying to complain, sorry if it sounded that way.

No worries, it didn't sound like that at all. I just know that I have been giving you a very large number of comments, and it has indeed been a long process. Hopefully you haven't been completely dissuaded from contributing in the future.

>  I'm much happier with this PR now than initially;

Agree with you there.

> let's merge it.

Does the button work for you, or do I need to do it?

minimaximalist commented:
I'll definitely continue to contribute, this package is great. I don't have write access, so I can't merge this PR.
